skrifa/src/meta/charmap.rs

```diff
@@ -0,0 +1,374 @@
+//! Mapping characters to nominal glyph identifiers.
+//!
+//! The functionality in this module provides a 1-to-1 mapping from Unicode
+//! characters (or [Unicode variation sequences](http://unicode.org/faq/vs.html)) to
+//! nominal or "default" internal glyph identifiers for a given font.
+//! This is a necessary first step, but generally insufficient for proper layout of
+//! [complex text](https://en.wikipedia.org/wiki/Complex_text_layout) or even
+//! simple text containing diacritics and ligatures.
+//!
+//! Comprehensive mapping of characters to positioned glyphs requires a process called
+//! shaping. For more detail, see: [Why do I need a shaping engine?](https://harfbuzz.github.io/why-do-i-need-a-shaping-engine.html)
+
+use read_fonts::{
+    tables::cmap::{self, Cmap, Cmap12, Cmap14, Cmap4, CmapSubtable, EncodingRecord, PlatformId},
+    types::{GlyphId, Uint24},
+    FontData, TableProvider,
+};
+
+pub use read_fonts::tables::cmap::MapVariant;
+
+/// Mapping of characters to nominal glyph identifiers.
+///
+/// The mappings are derived from the [cmap](https://learn.microsoft.com/en-us/typography/opentype/spec/cmap)
+/// table.
+///
+/// ## Supported formats
```

Comment:
Great question. My understanding is that the non-Unicode tables are simply unsupported. While both FreeType and HarfBuzz (and read-fonts) are capable of parsing them, they are never selected unless specifically requested because all inputs are expected to be Unicode and none of these libraries contain internal converters for legacy encodings. I’ll investigate this some more.

Reply:
Based on this: https://skia.googlesource.com/skia/+/c7b7b8fff26925debf40d30ae79b79b959fc91c3/src/ports/SkFontHost_FreeType.cpp#403, it looks like Skia accepts the default Unicode mapping or falls back to symbol if no Unicode mapping exists. I don't see code for selecting any other encodings which leads me to believe that we won't suffer regressions with this PR.

The fallback order does imply that we are doing the opposite here with regard to the preference for a symbol encoding. Since HarfBuzz does prioritize symbol, and per our chat last week, we discovered that the HB code is used for character mapping in Blink, I'm inclined to keep the strategy here as is.

Thoughts?

Reply:
Sorry for breaking threading, commented here: https://github.com/googlefonts/fontations/pull/316/files#r1157276977